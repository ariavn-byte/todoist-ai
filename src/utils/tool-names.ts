/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
