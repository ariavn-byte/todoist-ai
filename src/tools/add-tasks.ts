import type { AddTaskArgs, Task, TodoistApi } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { mapTask } from '../tool-helpers.js'
import { DurationParseError, parseDuration } from '../utils/duration-parser.js'
import {
    generateTaskNextSteps,
    getDateString,
    summarizeTaskOperation,
} from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const TaskSchema = z.object({
    content: z.string().min(1).describe('The content of the task to create.'),
    description: z.string().optional().describe('The description of the task.'),
    priority: z.number().int().min(1).max(4).optional().describe('The priority of the task (1-4).'),
    dueString: z.string().optional().describe('The due date for the task, in natural language.'),
    duration: z
        .string()
        .optional()
        .describe(
            'The duration of the task. Use format: "2h" (hours), "90m" (minutes), "2h30m" (combined), or "1.5h" (decimal hours). Max 24h.',
        ),
    projectId: z.string().optional().describe('The project ID to add this task to.'),
    sectionId: z.string().optional().describe('The section ID to add this task to.'),
    parentId: z.string().optional().describe('The parent task ID (for subtasks).'),
})

const ArgsSchema = {
    tasks: z.array(TaskSchema).min(1).describe('The array of tasks to add.'),
}

const addTasks = {
    name: ToolNames.ADD_TASKS,
    description: 'Add one or more tasks to a project, section, or parent.',
    parameters: ArgsSchema,
    async execute({ tasks }, client) {
        const addTaskPromises = tasks.map((task) => processTask(task, client))
        const newTasks = await Promise.all(addTaskPromises)
        const mappedTasks = newTasks.map(mapTask)

        const textContent = generateTextContent({
            tasks: mappedTasks,
            args: { tasks },
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks: mappedTasks,
                totalCount: mappedTasks.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

async function processTask(task: z.infer<typeof TaskSchema>, client: TodoistApi): Promise<Task> {
    const { duration: durationStr, projectId, sectionId, parentId, ...otherTaskArgs } = task

    let taskArgs: AddTaskArgs = { ...otherTaskArgs, projectId, sectionId, parentId }

    // Parse duration if provided
    if (durationStr) {
        try {
            const { minutes } = parseDuration(durationStr)
            taskArgs = {
                ...taskArgs,
                duration: minutes,
                durationUnit: 'minute',
            }
        } catch (error) {
            if (error instanceof DurationParseError) {
                throw new Error(`Task "${task.content}": ${error.message}`)
            }
            throw error
        }
    }

    return await client.addTask(taskArgs)
}

function generateTextContent({
    tasks,
    args,
}: {
    tasks: ReturnType<typeof mapTask>[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
}) {
    // Get context for smart next steps
    const todayStr = getDateString()
    const hasToday = tasks.some((task) => task.dueDate === todayStr)

    // Generate context description for mixed contexts
    const contextTypes = new Set<string>()
    for (const task of args.tasks) {
        if (task.projectId) contextTypes.add('projects')
        else if (task.sectionId) contextTypes.add('sections')
        else if (task.parentId) contextTypes.add('subtasks')
        else contextTypes.add('inbox')
    }

    let projectContext = ''
    if (contextTypes.size === 1) {
        const contextType = Array.from(contextTypes)[0]
        projectContext = contextType === 'inbox' ? '' : `to ${contextType}`
    } else if (contextTypes.size > 1) {
        projectContext = 'to multiple contexts'
    }

    return summarizeTaskOperation('Added', tasks, {
        context: projectContext,
        nextSteps: generateTaskNextSteps('added', tasks, { hasToday }),
        showDetails: true,
    })
}

export { addTasks }
