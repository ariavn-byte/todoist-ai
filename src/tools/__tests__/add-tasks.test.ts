import type { Task, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    TODAY,
    createMockTask,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { addTasks } from '../add-tasks.js'

// Mock the Todoist API
const mockTodoistApi = {
    addTask: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { ADD_TASKS, GET_OVERVIEW } = ToolNames

describe(`${ADD_TASKS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('adding multiple tasks', () => {
        it('should add multiple tasks and return mapped results', async () => {
            // Mock API responses extracted from recordings (Task type)
            const mockApiResponse1: Task = createMockTask({
                id: '8485093748',
                content: 'First task content',
                url: 'https://todoist.com/showTask?id=8485093748',
                addedAt: '2025-08-13T22:09:56.123456Z',
            })

            const mockApiResponse2: Task = createMockTask({
                id: '8485093749',
                content: 'Second task content',
                description: 'Task description',
                labels: ['work', 'urgent'],
                childOrder: 2,
                priority: 2,
                url: 'https://todoist.com/showTask?id=8485093749',
                addedAt: '2025-08-13T22:09:57.123456Z',
                due: {
                    date: '2025-08-15',
                    isRecurring: false,
                    lang: 'en',
                    string: 'Aug 15',
                    timezone: null,
                },
            })

            mockTodoistApi.addTask
                .mockResolvedValueOnce(mockApiResponse1)
                .mockResolvedValueOnce(mockApiResponse2)

            const result = await addTasks.execute(
                {
                    tasks: [
                        { content: 'First task content', projectId: '6cfCcrrCFg2xP94Q' },
                        {
                            content: 'Second task content',
                            description: 'Task description',
                            priority: 2,
                            dueString: 'Aug 15',
                            projectId: '6cfCcrrCFg2xP94Q',
                        },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each task
            expect(mockTodoistApi.addTask).toHaveBeenCalledTimes(2)
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(1, {
                content: 'First task content',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
            })
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(2, {
                content: 'Second task content',
                description: 'Task description',
                priority: 2,
                dueString: 'Aug 15',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.tasks).toHaveLength(2)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    totalCount: 2,
                    tasks: expect.arrayContaining([
                        expect.objectContaining({ id: '8485093748' }),
                        expect.objectContaining({ id: '8485093749' }),
                    ]),
                }),
            )
        })

        it('should handle tasks with section and parent IDs', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093750',
                content: 'Subtask content',
                description: 'Subtask description',
                priority: 3,
                sectionId: 'section-123',
                parentId: 'parent-task-456',
                url: 'https://todoist.com/showTask?id=8485093750',
                addedAt: '2025-08-13T22:09:58.123456Z',
            })

            mockTodoistApi.addTask.mockResolvedValue(mockApiResponse)

            const result = await addTasks.execute(
                {
                    tasks: [
                        {
                            content: 'Subtask content',
                            description: 'Subtask description',
                            priority: 3,
                            projectId: '6cfCcrrCFg2xP94Q',
                            sectionId: 'section-123',
                            parentId: 'parent-task-456',
                        },
                    ],
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addTask).toHaveBeenCalledWith({
                content: 'Subtask content',
                description: 'Subtask description',
                priority: 3,
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: 'section-123',
                parentId: 'parent-task-456',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    totalCount: 1,
                    tasks: expect.arrayContaining([expect.objectContaining({ id: '8485093750' })]),
                }),
            )
        })

        it('should add tasks with duration', async () => {
            const mockApiResponse1: Task = createMockTask({
                id: '8485093752',
                content: 'Task with 2 hour duration',
                duration: { amount: 120, unit: 'minute' },
                url: 'https://todoist.com/showTask?id=8485093752',
                addedAt: '2025-08-13T22:09:56.123456Z',
            })

            const mockApiResponse2: Task = createMockTask({
                id: '8485093753',
                content: 'Task with 45 minute duration',
                duration: { amount: 45, unit: 'minute' },
                url: 'https://todoist.com/showTask?id=8485093753',
                addedAt: '2025-08-13T22:09:57.123456Z',
            })

            mockTodoistApi.addTask
                .mockResolvedValueOnce(mockApiResponse1)
                .mockResolvedValueOnce(mockApiResponse2)

            const result = await addTasks.execute(
                {
                    tasks: [
                        {
                            content: 'Task with 2 hour duration',
                            duration: '2h',
                            projectId: '6cfCcrrCFg2xP94Q',
                        },
                        {
                            content: 'Task with 45 minute duration',
                            duration: '45m',
                            projectId: '6cfCcrrCFg2xP94Q',
                        },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called with parsed duration
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(1, {
                content: 'Task with 2 hour duration',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
                duration: 120,
                durationUnit: 'minute',
            })
            expect(mockTodoistApi.addTask).toHaveBeenNthCalledWith(2, {
                content: 'Task with 45 minute duration',
                projectId: '6cfCcrrCFg2xP94Q',
                sectionId: undefined,
                parentId: undefined,
                duration: 45,
                durationUnit: 'minute',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.tasks).toHaveLength(2)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    totalCount: 2,
                    tasks: expect.arrayContaining([
                        expect.objectContaining({ id: '8485093752' }),
                        expect.objectContaining({ id: '8485093753' }),
                    ]),
                }),
            )
        })

        it('should handle various duration formats', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093754',
                content: 'Task with combined duration',
                duration: { amount: 150, unit: 'minute' },
                url: 'https://todoist.com/showTask?id=8485093754',
                addedAt: '2025-08-13T22:09:56.123456Z',
            })

            mockTodoistApi.addTask.mockResolvedValue(mockApiResponse)

            // Test different duration formats
            const testCases = [
                { input: '2h30m', expectedMinutes: 150 },
                { input: '1.5h', expectedMinutes: 90 },
                { input: ' 90m ', expectedMinutes: 90 },
                { input: '2H30M', expectedMinutes: 150 },
            ]

            for (const testCase of testCases) {
                mockTodoistApi.addTask.mockClear()

                await addTasks.execute(
                    { tasks: [{ content: 'Test task', duration: testCase.input }] },
                    mockTodoistApi,
                )

                expect(mockTodoistApi.addTask).toHaveBeenCalledWith(
                    expect.objectContaining({
                        duration: testCase.expectedMinutes,
                        durationUnit: 'minute',
                    }),
                )
            }
        })
    })

    describe('error handling', () => {
        it('should throw error for invalid duration format', async () => {
            await expect(
                addTasks.execute(
                    { tasks: [{ content: 'Task with invalid duration', duration: 'invalid' }] },
                    mockTodoistApi,
                ),
            ).rejects.toThrow(
                'Task "Task with invalid duration": Invalid duration format "invalid"',
            )
        })

        it('should throw error for duration exceeding 24 hours', async () => {
            await expect(
                addTasks.execute(
                    { tasks: [{ content: 'Task with too long duration', duration: '25h' }] },
                    mockTodoistApi,
                ),
            ).rejects.toThrow(
                'Task "Task with too long duration": Invalid duration format "25h": Duration cannot exceed 24 hours (1440 minutes)',
            )
        })

        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Task content is required')
            mockTodoistApi.addTask.mockRejectedValue(apiError)

            await expect(
                addTasks.execute({ tasks: [{ content: '' }] }, mockTodoistApi),
            ).rejects.toThrow(apiError.message)
        })

        it('should handle partial failures when adding multiple tasks', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093751',
                content: 'First task content',
                url: 'https://todoist.com/showTask?id=8485093751',
                addedAt: '2025-08-13T22:09:59.123456Z',
            })

            const apiError = new Error('API Error: Second task failed')
            mockTodoistApi.addTask
                .mockResolvedValueOnce(mockApiResponse)
                .mockRejectedValueOnce(apiError)

            await expect(
                addTasks.execute(
                    {
                        tasks: [
                            { content: 'First task content' },
                            { content: 'Second task content' },
                        ],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Second task failed')

            // Verify first task was attempted
            expect(mockTodoistApi.addTask).toHaveBeenCalledTimes(2)
        })
    })

    describe('next steps logic', () => {
        it('should suggest find-tasks-by-date for today when hasToday is true', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093755',
                content: 'Task due today',
                url: 'https://todoist.com/showTask?id=8485093755',
                addedAt: '2025-08-13T22:09:56.123456Z',
                due: {
                    date: TODAY,
                    isRecurring: false,
                    lang: 'en',
                    string: 'today',
                    timezone: null,
                },
            })

            mockTodoistApi.addTask.mockResolvedValue(mockApiResponse)

            const result = await addTasks.execute(
                { tasks: [{ content: 'Task due today', dueString: 'today' }] },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                `Use ${GET_OVERVIEW} to see your updated project organization`,
            )
        })

        it('should suggest overview tool when no hasToday context', async () => {
            const mockApiResponse: Task = createMockTask({
                id: '8485093756',
                content: 'Regular task',
                url: 'https://todoist.com/showTask?id=8485093756',
                addedAt: '2025-08-13T22:09:56.123456Z',
            })

            mockTodoistApi.addTask.mockResolvedValue(mockApiResponse)

            const result = await addTasks.execute(
                {
                    tasks: [{ content: 'Regular task', projectId: '6cfCcrrCFg2xP94Q' }],
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                `Use ${GET_OVERVIEW} to see your updated project organization`,
            )
        })
    })
})
