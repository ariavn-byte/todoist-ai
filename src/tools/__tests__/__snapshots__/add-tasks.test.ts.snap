// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`add-tasks tool adding multiple tasks should add multiple tasks and return mapped results 1`] = `
"Added 2 tasks to projects.
Tasks:
    First task content • P1 • id=8485093748
    Second task content • due 2025-08-15 • P2 • id=8485093749.
Next:
- Use get-overview to see your updated project organization"
`;

exports[`add-tasks tool adding multiple tasks should add tasks with duration 1`] = `
"Added 2 tasks to projects.
Tasks:
    Task with 2 hour duration • P1 • id=8485093752
    Task with 45 minute duration • P1 • id=8485093753.
Next:
- Use get-overview to see your updated project organization"
`;

exports[`add-tasks tool adding multiple tasks should handle tasks with section and parent IDs 1`] = `
"Added 1 task to projects.
Tasks:
    Subtask content • P3 • id=8485093750.
Next:
- Use get-overview to see your updated project organization"
`;

exports[`add-tasks tool next steps logic should suggest find-tasks-by-date for today when hasToday is true 1`] = `
"Added 1 task.
Tasks:
    Task due today • due 2025-08-17 • P1 • id=8485093755.
Next:
- Use get-overview to see your updated project organization"
`;

exports[`add-tasks tool next steps logic should suggest overview tool when no hasToday context 1`] = `
"Added 1 task to projects.
Tasks:
    Regular task • P1 • id=8485093756.
Next:
- Use get-overview to see your updated project organization"
`;
